### ISSUE TITLE
evaluate if a state machine is a good abstraction for tracking connection state

### ISSUE THREAD

**Issue by alexanderkjall** [pgjdbc/pgsql2#5]
I have started on having a state machine in the ProtocolV3 class, with a set of allowed state transitions in ProtocolV3States. And I'm not totally sure if this is a good path to take in order to track the connection and what happens in it.

The state transition table might get very complex, specially if all authentication methods and tls things should go into it.

**Comment by davecramer** [comment]
The protocol isn't that complicated. Have you looked at the current implementation ?

**Comment by alexanderkjall** [comment]
I agree, the base things are really straight forward. I think I was a bit intimidated by all the different authentication methods and tls. It felt like the number of different combinations could get pretty high.

I think I'll just implement md5 for now, and instead focus on getting queries to work.

**Comment by davecramer** [comment]
FYI https://github.com/oracle/oracle-db-examples/tree/master/java/AoJ

**Comment by alexanderkjall** [comment]
That is really interesting, this will help a lot with understanding the new standard.

I notice that the adba packages are located in jdk.incubator.sql2, maybe we should move our packages to reflect that.

**Comment by alexanderkjall** [comment]
yes, I meant that we should rename the java2.sql2 catalog to jdk.incubator.sql2, so that it matches the sandbox. I just pushed a commit that did that and updated the code to match the latest interface changes.
